child-device requirements (#1195)
thin-edge/thin-edge.io: docs/src/requirements/child-devices.md

# Child Devices

## Child Device Model
* A child device implies two aspects:
  - a "physical child-device" on device site
  - a "cloud child device-twin" (a logical representation of the physical device in the cloud)
  - ?an edge child device-twin (a logical representation of the physical device in thin-edge)
* Physical child devices typically are special-purpose hardware devices designed to do just a single task or a handful of tasks at best. 
* These devices can't (or don't want, out of security reasons) connect to the cloud directly. Instead they get connected to the cloud via a gateway device.
* The gateway device runs thin-edge, and all physical child devices are connected to the gateway device via network, fieldbus or any other channel. 

> * The gateway device appears in the cloud with it's digital device-twin, with one digital device-twin per each physical child-device underneath.

Reviewer: The mapping is not always 1:1. There could be multiple physical devices grouped together and represented as a single child device. Similarly, a physical device connected to the cloud may not want a separate device twin in the cloud but just associate itself to the device twin of the gateway itself.

Author: Good point. Even a "child device-twin" in the cloud is possible without any "physical child-device" (e.g. when just a local process on the gateway uses the child-device representation).

We might need to put some effort to really explicit define scope and difference of terms "child device-twin", "physical child-device" and the general term "child-device".

* The relation between a physical child-device and the according cloud device-twin is made with a child-id.
* The gateway takes care of routing all data/requests/responses beetween the physical child-devices and the related cloud child-device twins.
* Multiple physical child devices could be connected to a single gateway.

TODO: do not forget, a child-device could be also logical (a process running on the gateway)

# User Journey

## Connecting a smart physical child-device (featuring MQTT client) to the cloud via the gateway

> - User creates the child-device twin in the cloud, with a unique child-id and supported operations

Reviewer: Define which type of user? Device admin? Device app developer?

# Child Device Concept

TODO: Add motivation for child-device concept. Somethin like:<br/>
-- Device-side and Cloud-side structural view (that might match the physical structure)<br/>
-- Features are available per child-device (e.g. Software Management with one SW list per child device, ...)

That document summarises a collection of use-cases and requirements, based on experiences of Sofware AG and some customers/partners.

## Definition
1) A child device is a physical device connected to the thin-edge device (typically a gateway device)  or even over IP.
2) Generally there are two kinds of child devices:
   - a) A dumb device that is connected to the thin-edge device via an IO-Link or fieldbus, that can't run the processing logic that sends/receives data to/from the thin-edge device. For such devices, the processing logic will be running on the thin-edge device itself.
   - b) A smart device that is connected to the thin-edge device over an IP network, with the processing logic running on that device itself. Its interactions with the thin-edge device will only be over the network

> 3) In both kinds of devices, there will always be a "child device process" that is responsible for handling all data to/from the physical child device.

Reviewer: ```suggestion
3) In both kinds of devices, there will always be a "child device process" that is responsible for handling all data to/from the physical child device. This process will be developed by thin-edge.io users against thin-edge APIs.
```

> 4) Thin-edge's responsibility is to receive and route any cloud-bound data from these child device processes and any device-bound data like software/config binaries to these child device processes.

Reviewer: ```suggestion
4) Thin-edge's responsibility is to receive and route any cloud-bound data from these child device processes to all the connected clouds and any device-bound data from the cloud, like software/config binaries, to these child device processes.
```

> 5) The child device process itself is resposible for pushing these binaries to the physical device itself and applying it there.

Reviewer: ```suggestion
5) The child device process itself is responsible for pushing these binaries to the physical device itself and applying it there.
```

6) Each child device has a unique ID that can be used to identify/distinguish between them

> TODO: Add somehow statement below as definition: 
> "PLCs are already long time in production-line and no additional software can be installed (e.g. MQTT client) or configuration changed, to make the PLC capbable to communicate with the thin-edge device. That is to avoid any risc in the production line. Instead the thin-edge device must use existing interfaces from the PLC."

Reviewer: ```suggestion
"On PLCs that are already running in production-line, no new software can be installed (e.g. an MQTT client, or a new C application) or configurations changed, to make the PLC communicate with the thin-edge device directly. Instead, the thin-edge device must use existing interfaces from the PLC to interact with it. That is to avoid any risks in the production line."
```

## Questions
1) Are child devices and their unique IDs static and known at thin-edge provisioning time, or do they dynamically get connected to thin-edge?
2) Should thin-edge provide routing of data between child devices (from one child device to another)?
3) Will there be one "child device process" per "physical child device"? Or a single process can manage multiple physical child devices at once?
4) Should thin-edge also support receiving data from multiple child devices and combining all that data and associate it with the main thin-edge device twin itself in the cloud(instead of having multiple child devices)? Customer typically do this as a cost-cutting measure with clouds charging fees per device.

## Use-Cases

* **Use-Case 1: Gathering telemetry data from connected PLCs and handle as Measurements, Alarms or Events**
   - a process on the thin-edge device collects process-data from a CODESYS PLC via local network (using the CODESYS PLC handler)

>    - per each connected PLC a process is running

Reviewer: ```suggestion
   - once such process is running per connected PLC
```

>    - the process pushes gathered PLC's process-data to thin-edge measurements, alarms or events

Reviewer: ```suggestion
   - this process pushes the process-data gathered from these PLCs as thin-edge JSON measurements, alarms or events.
```

>    - the process has a configuration file that contains mapping information, meaning which process-data to push to which thin-edge measurement, alarm or event;

Reviewer: ```suggestion
   - the process has a data mapping configuration file that defines which PLC process-data is to be mapped into a thin-edge measurement, alarm or event
```

     and which child-device ID to use
   

* **Use-Case 2: Viewing/applying data-mapping configuration of PLCs using Configuration Management**

>    - the process' configuration (see process configuration file in use-case 1) is managed by Configuration Management

Reviewer: ```suggestion
   - the process' data mapping configuration file (see process configuration file in use-case 1) is managed by the Configuration Management feature
```

   - each PLC appears as child-device in the Cloud, that supports individual Configuration Management functionality per child-device

>    - thereby data mapping can be changed or extended from cloud side

Reviewer: ```suggestion
   - this enables the data mapping to be changed or extended from cloud
```

* **Use-Case 3: Update application on connected PLCs using Software Management**
   - each PLC appears as child-device in the Cloud, that supports individual Software Management functionality    

>    - a commandline tool on the thin-edge device is capable to send a CODESYS application (one file) to a connected PLC, and start it

Reviewer: ```suggestion
   - a command-line tool on the thin-edge device is capable of sending a CODESYS application (one file) to a connected PLC, and start it
```

>    - on an incoming SW Management update request for a PLC that commandline tool is invoked to install the CODESYS application on the according PLC

Reviewer: ```suggestion
   - on an incoming SW Management update request for a PLC, that command-line tool must be invoked to install the CODESYS application on the according PLC
```

NOTE 1: For use-cases above the external device could be also another PLC than CODESYS or even another device than a PLC.

NOTE 2: Use-case(s) with external device that can directly access thin-edge interfaces (e.g. tedge/measurements) not yet covered.

# Requirements

- **Requirement 1:** Thin-edge has to provide child-device support for interfaces: <br/>
     - Measurements/Alarm/Events<br/>
     - Configuration Managements<br/>

>      - Software Managements Plugin API<br/>

Reviewer: ```suggestion
     - Software Management
```

> - **Requirement 2:** Where thin-edge provides child-device support it uses corresponding Clouds' child device meachnism, to associate data/functionality of that child-device as own device underneath the thin-edge device.

Reviewer: ```suggestion
- **Requirement 2:** Each child device that is connected to the thin-edge device must have a corresponding device twin on the cloud as well, appearing under the parent thin-edge device as child devices. All the data coming from these physical child devices must be associated with their respective device twins in the cloud.
```

> - **Requirement 3:** When data is pushed to thin-edge for a child-device, thin-edge will take care that the child-device exists in the Cloud, and creates it if needed.

Reviewer: ```suggestion
- **Requirement 3:** When data is pushed to thin-edge for a child-device, thin-edge will take care of creating that child device twin in the Cloud, if it doesn't exist already.
```

> - **Requirement 3:** When data is pushed to thin-edge for a child-device, thin-edge will take care that the child-device exists in the Cloud, and creates it if needed.

Reviewer: Not fully convinced of this requirement yet. It adds unnecessary burden on thin-edge to be stateful(to maintain the list of child devices created already)

- **Requirement 4:** Thin-edge is stateless in terms of child devices. I.e. it does not store a list of child-devices of even a state of those. All knowledge about child-devices is in the applications that provide/consume data and functionality from/to child-devices. 

> - Requrirement 5: thin-edge fetches the Configuration from the device and pushed to the cloud

Reviewer: ```suggestion
- **Requirement 5:** thin-edge can fetch configurations from the device and push to the cloud
```

    - Thin-edge should deliver the config upload request from the cloud to the appropriate child device process
    - Expose an API with which the child device process can upload the snapshot of the config file binary to the thin-edge device.
    - Thin-edge should then upload and associate these config files with their corresponding device twin in the cloud.
    - Thin-edge must make sure that config files uploaded by one child device should not overwrite the config uploaded by another child, even if they are of the same type

> - Requrirement 6: thin-edge downloads configuration file from the cloud and delivers the configuration to the child-device process

Reviewer: ```suggestion
- **Requirement 6**: thin-edge can download configuration files from the cloud and delivers that file to the child-device process
```

TODO: do not forget, a child-device could be also logical (a process running on the gateway)

# Initialisation of a Child-Device
Before a child-device could be put into action, both aspects from above (_Device-Side Endpoint_ and _Cloud Child Device-Twin_) need to be initialised.

* Initialisation: Device-Side Endpoint

   ...TODO...

* Initialisation: Cloud Child Device-Twin

  Two levels of initialisations have to take place for the cloud child device-twin: 
  1) creating the new child-device twin in the cloud
  2) declaring capabilities on the new child-device twin, i.E.:
     - declaring supported operations (e.g. sw-update, config-man, log-man, ...)
     - declaring supported types (e.g. sw module-type, config-type, log-type, ...)

# User Journey

> ## Definition of Personas used in the user journeys below

Reviewer 2: I would not use the concept of *persona* to which I prefer the concept of *user role* - at least for this kind of design work. A user role is focused on the user's goals. While a persona might be extended to distinguish abilities and interests.

* __Device Operator__ (the one who maintains/operates the device in the field in the shopfloor)

> * __Agent Developer__ (the one who adapts and deploys thin-edge to the device)
> * __Device App Developer__ (the one who implements the device application that makes use of thin-edge APIs)

Reviewer 2: What difference do you make between these two roles?

I see no differences and prefer the first name and the second description ;-)

```suggestion
* __Agent Developer__ (the one who implements the device application that makes use of thin-edge APIs)
```

> * __Solution Architect__ (the one, who is in authority for the whole end-to-end picture (from devices to cloud) )

Reviewer 2: I would simply ignore this role. Thin-edge provides nothing for these architects beyond APIs and tools to be respectively used by agent developers and device operators.

## Journey: Connecting a physical child-device to the cloud via the gateway

Preface: There are two options to create a child-device twin in the cloud:
  1) by cloud site or by some customer specific tool/app on-top of the cloud
  2) by the thin-edge device<br/>

>      a) automatically no 1st data push<br/>

Reviewer 2: Do you meant:

```suggestion
     a) automatically on 1st data push<br/>
```

     b) intentionally by configuration or external device request

Same two options are possible for declaring capabilities.

> Scenarios below are per each combination of those options:

Reviewer 2: Having a scenario per combination will inevitably lead to copy/paste, nobody reading the scenarios and stupid bugs. Please don't.
